### ISSUE TITLE
Add support for prettier 3.x

### ISSUE THREAD

**Issue by DavidGregory084** [diffplug/spotless#1751]
I'm trying to update the version of prettier used in [a VS Code plugin](https://codeberg.org/mina-lang/mina/src/branch/main/vscode-plugin/build.gradle.kts).

Currently I'm using the default version of prettier that's installed by spotless - this seems to be version 2.8.1.

If I update my spotless configuration to use version 3.0.0 and use the workaround mentioned in #1750 to get spotless installed, then Spotless replaces my TypeScript code with an empty JS object.

Gradle version: 8.2.1
Spotless version: 6.19.0

Spotless configuration:

```
spotless {
    ratchetFrom("origin/main")

    typescript {
        target("src/**/*.ts")
        prettier("3.0.0")
    }

    json {
        target("syntaxes/**/*.json")
        prettier("3.0.0")
    }
}
```

`build` command output:
```
* What went wrong:
Execution failed for task ':vscode-plugin:spotlessTypescriptCheck'.
> The following files had format violations:
      src/extension.ts
          @@ -1,273 +1 @@
          -import·*·as·child_process·from·"child_process";
          -import·*·as·os·from·"os";
          -import·*·as·path·from·"path";
          -import·*·as·util·from·"util";
          -
          -import·{
          -··ExtensionContext,
          -··OutputChannel,
          -··ProgressLocation,
          -··Uri,
          -··window,
          -··workspace,
          -}·from·"vscode";
          -
          -import·{
          -··LanguageClient,
          -··LanguageClientOptions,
          -··ServerOptions,
          -··TransportKind,
          -}·from·"vscode-languageclient/node";
          -
          -import·{·getApi,·FileDownloader·}·from·"@microsoft/vscode-file-downloader-api";
          -import·{·findRuntimes·}·from·"jdk-utils";
          -
          -const·COURSIER_DOWNLOAD_TIMEOUT_MS·=·120000;
          -const·COURSIER_LAUNCHER_URI·=·Uri.parse(
          -··"https://github.com/coursier/launchers/raw/master/coursier.jar",
          -);
          -
          -let·client:·LanguageClient;
          -
          -async·function·downloadCoursierJar(
          -··context:·ExtensionContext,
          -··fileDownloader:·FileDownloader,
          -):·Promise<Uri>·{
          -··return·await·window.withProgress(
          -····{
          -······location:·ProgressLocation.Window,
          -····},
          -····(progress,·cancellationToken)·=>·{
          -······return·fileDownloader.downloadFile(
          -········COURSIER_LAUNCHER_URI,
          -········"coursier.jar",
          -········context,
          -········cancellationToken,
          -········(_downloadedBytes,·_totalBytes)·=>
          -··········progress.report({·message:·"Downloading·coursier·JAR"·}),
          -········{·timeoutInMs:·COURSIER_DOWNLOAD_TIMEOUT_MS·},
      ... (226 more lines that didn't fit)
  Run './gradlew :vscode-plugin:spotlessApply' to fix these violations.
```

If I run `./gradlew :vscode-plugin:spotlessApply` as suggested it applies [this diff](https://gist.github.com/DavidGregory084/38778f2b08315fc570959aff6f6e29f9) to the file.

If I run `npx prettier --write src` from within the `vscode-plugin` folder to use prettier directly, I don't get the same result, so I don't think it's an issue in prettier itself.

**Comment by nedtwigg** [comment]
Looks like `prettier 3.0` converted all [api methods to return Promise](https://prettier.io/blog/2023/07/05/3.0.0.html#api-1). Adapting to this is a decent amount of work. @simschla might have some pointers for how to go about this.

**Comment by simschla** [comment]
Yes, this looks like it needs some work to get it to be compatible with both 2.x and 3.x API. 
I'll gladly look into it. 
(Might be 2–3 weeks until I'm able to dive into it due to upcoming vacations, I apologize for any inconvenience this is causing.)

**Comment by DavidGregory084** [comment]
No inconvenience at all, this was a speculative upgrade - enjoy your time off! 🏖️

**Comment by nedtwigg** [comment]
Released in `plugin-gradle 6.21.0` and `plugin-maven 2.38.0`.
